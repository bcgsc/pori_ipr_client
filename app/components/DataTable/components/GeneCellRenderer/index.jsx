import React, { useState } from 'react';
import GeneViewer from '../GeneViewer';

import './index.scss';

/**
 * @param {object} params params
 * @param {string} params.value display text
 * @return {*} JSX
 */
function GeneCellRenderer(params) {
  const {
    value,
    context: { reportIdent },
  } = params;

  const [showGeneViewer, setShowGeneViewer] = useState(false);

  return (
    <>
      {value && value.split(/\s::\s|,\s/).map((val, index) => (
        <React.Fragment key={index}>
          {index > 0 && (
            <span>
              {value.includes(' :: ') ? ' :: ' : ', '}
            </span>
          )}
          <span
            tabIndex={0}
            role="button"
            onClick={() => setShowGeneViewer(val)}
            onKeyDown={() => setShowGeneViewer(val)}
            className="gene__text"
          >
            {val}
          </span>
          <>
            {showGeneViewer && (
              <GeneViewer
                open={showGeneViewer === val}
                gene={val}
                reportIdent={reportIdent}
                onClose={() => setShowGeneViewer(false)}
              />
            )}
          </>
        </React.Fragment>
      ))}
    </>
  );
}

export default GeneCellRenderer;
