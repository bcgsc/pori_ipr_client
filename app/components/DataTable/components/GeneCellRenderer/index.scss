@import '../../../../styles/variables';

.gene {
  &__text {
    color: $underline;
    font-weight: 700;
    text-decoration: underline solid $underline;
    cursor: pointer;
  }
}