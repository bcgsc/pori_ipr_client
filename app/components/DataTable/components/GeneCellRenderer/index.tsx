import React, { useState } from 'react';

import GeneViewer from '../GeneViewer';

import './index.scss';

type GeneCellRendererProps = {
  value: string;
  link?: boolean;
};

const GeneCellRenderer = ({
  value,
  link = false,
}: GeneCellRendererProps) => (
  value && value.split(/\s*::\s*|,\s?/).map((val, index) => (
    // eslint-disable-next-line react/no-array-index-key
    <React.Fragment key={`${val}_${index}`}>
      {index > 0 && (
        <span>
          {value.includes(' :: ') ? ' :: ' : ', '}
        </span>
      )}
      <GeneViewer
        isLink={link}
        gene={val}
      />
    </React.Fragment>
  ))

);

export default GeneCellRenderer;
