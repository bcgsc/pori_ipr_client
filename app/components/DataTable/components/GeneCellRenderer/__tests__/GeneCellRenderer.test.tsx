import React from 'react';
import { screen, render, fireEvent } from '@testing-library/react';

import GeneCellRenderer from '..';

const mockGene = 'EGFR';
const mockFusionCommas = 'CABLES1,DCC';
const mockFusionColons = 'CABLES1 :: DCC';
const expectedFusions = ['CABLES1', 'DCC'];

// eslint-disable-next-line react/display-name
jest.mock('../../GeneViewer', () => () => (<div role="presentation" />));

describe('GeneCellRenderer', () => {
  test('It matches the snapshot', () => {
    const { asFragment } = render(
      <GeneCellRenderer
        value={mockGene}
        link
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('It renders the gene and the link button', async () => {
    render(
      <GeneCellRenderer
        value={mockGene}
        link
      />,
    );

    expect(await screen.findByText(mockGene)).toBeInTheDocument();
    expect(await screen.findByRole('button')).toBeInTheDocument();
  });

  test('It renders the gene without a link button', async () => {
    render(
      <GeneCellRenderer
        value={mockGene}
        link={false}
      />,
    );

    expect(await screen.findByText(mockGene)).toBeInTheDocument();
    expect(screen.queryByRole('button')).not.toBeInTheDocument();
  });

  test('It renders a fusion gene with a comma', async () => {
    render(
      <GeneCellRenderer
        value={mockFusionCommas}
        link={false}
      />,
    );

    expectedFusions.forEach((fusion) => {
      expect(screen.getByText(fusion)).toBeInTheDocument();
    });
    expect(screen.getByText(',')).toBeInTheDocument();
  });

  test('It renders a fusion gene with colons', async () => {
    render(
      <GeneCellRenderer
        value={mockFusionColons}
        link={false}
      />,
    );

    expectedFusions.forEach((fusion) => {
      expect(screen.getByText(fusion)).toBeInTheDocument();
    });
    expect(screen.getByText('::')).toBeInTheDocument();
  });

  test('The GeneViewer is shown upon link button click', async () => {
    render(
      <GeneCellRenderer
        value={mockGene}
        link
      />,
    );

    fireEvent.click(await screen.findByRole('button'));
    expect(await screen.findByRole('presentation')).toBeInTheDocument();
  });
});
