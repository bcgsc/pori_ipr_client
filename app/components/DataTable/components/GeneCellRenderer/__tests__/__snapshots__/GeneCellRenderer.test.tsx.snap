// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneCellRenderer It matches the snapshot 1`] = `
<DocumentFragment>
  <span
    class="gene__text"
    role="button"
    tabindex="0"
  >
    EGFR
  </span>
</DocumentFragment>
`;
